### PR TITLE
integrate_ode_adams in Stan

### PR DESCRIPTION
**Pull request by yizhang-cae**
#### Submission Checklist

- [x] Run unit tests: `./runTests.py src/test/unit`
- [x] Run cpplint: `make cpplint`
- [x] Declare copyright holder and open-source license: see below

#### Summary
With https://github.com/stan-dev/math/issues/734 we are ready to provde Adams-Moulton integrator to Stan language. The process is identical to `rk45` and `bdf` integrators, as the function signature as well as ODE functor signature remain the same.

#### Intended Effect
Support `integrate_ode_adams` function as the third ODE solver in Stan.

#### How to Verify
Unit tests and model tests.

#### Side Effects
n/a

#### Copyright and Licensing
Metrum Research Group, LLC

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company):

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: src/stan/lang/grammars/semantic_actions_def.cpp [hunk 1]
```diff
@@ -1625,8 +1625,11 @@ namespace stan {
     void deprecated_integrate_ode::operator()(std::ostream& error_msgs)
       const {
       error_msgs << "Warning: the integrate_ode() function is deprecated"
-             << " in the Stan language; use integrate_ode_rk45() [non-stiff]"
-             << " or integrate_ode_bdf() [stiff] instead."
+             << " in the Stan language; use the following functions"
+             << " instead."
+             << "  integrate_ode_rk45() [non-stiff]"
+             << "  integrate_ode_adams() [non-stiff]"
```
Comment (bob-carpenter) [inline src/stan/lang/grammars/semantic_actions_def.cpp]: This is where it'd be nice to have a suggestion about which to use when.

File: src/test/test-models/good/integrate_ode_adams.stan [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+  y_hat = integrate_ode_adams(sho, y0_p, t0, ts, theta_p, x, x_int);
+
+  y_hat = integrate_ode_adams(sho, y0_d, t0, ts, theta_d, x, x_int, 1e-10, 1e-10, 1e8);
+  y_hat = integrate_ode_adams(sho, y0_d, t0, ts, theta_p, x, x_int, 1e-10, 1e-10, 1e8);
+  y_hat = integrate_ode_adams(sho, y0_p, t0, ts, theta_d, x, x_int, 1e-10, 1e-10, 1e8);
+  y_hat = integrate_ode_adams(sho, y0_p, t0, ts, theta_p, x, x_int, 1e-10, 1e-10, 1e8);
+}
+generated quantities {
+  real y_hat[T,2];
+  y_hat = integrate_ode_adams(sho, y0_d, t0, ts, theta_d, x, x_int);
```
Comment (bob-carpenter) [inline src/test/test-models/good/integrate_ode_adams.stan]: These tests are for instantiations of functions in the language, not for assignment.  

The data argument versions should stay in generated quantities---that tests that the result is `double`.  They can be removed from transformed parameters as that just tests that a `double` can be promoted to `var` under assignment, which is another module's responsibility.

File: src/test/unit/lang/parser/ode_test.cpp [hunk 3]
```diff
@@ -119,3 +119,49 @@ TEST(lang_parser, integrate_ode_bdf_control_bad) {
...
+  test_throws("ode/adams/bad_t_type",
+          "third argument to integrate_ode_adams must have type real or int");
+  test_throws("ode/adams/bad_ts_type",
+              "fourth argument to integrate_ode_adams must have type real[]");
+  test_throws("ode/adams/bad_theta_type",
+              "fifth argument to integrate_ode_adams must have type real[]");
+  test_throws("ode/adams/bad_x_type",
+              "sixth argument to integrate_ode_adams must have type real[]");
+  test_throws("ode/adams/bad_x_int_type",
+              "seventh argument to integrate_ode_adams must have type int[]");
```
Comment (bob-carpenter) [inline src/test/unit/lang/parser/ode_test.cpp]: Thanks for being careful in going through all these.

File: src/stan/lang/grammars/semantic_actions_def.cpp (hunk 1 above)
Comment (yizhang-cae) [reply src/stan/lang/grammars/semantic_actions_def.cpp]: Updated msg
```c++
       error_msgs << "Warning: the integrate_ode() function is deprecated"
             << " in the Stan language; use the following functions"
             << " instead.\n"
             << " integrate_ode_rk45() [explicit, order 5, for non-stiff problems]\n"
             << " integrate_ode_adams() [implicit, up to order 12, for non-stiff problems]\n"
             << " integrate_ode_bdf() [implicit, up to order 5, for stiff problems]."
             << std::endl;
```

File: src/test/test-models/good/integrate_ode_adams.stan (hunk 2 above)
Comment (yizhang-cae) [reply src/test/test-models/good/integrate_ode_adams.stan]: Removed data argument tests in transformed parameters.

**Comment by yizhang-cae** [comment]
Thanks @bob-carpenter.
Updated per review comments.

File: src/stan/lang/grammars/semantic_actions_def.cpp (hunk 1 above)
Comment (bob-carpenter) [reply src/stan/lang/grammars/semantic_actions_def.cpp]: Thanks---that's great.  I can use that in the manual then, too.

**Pull request merged**